tokio/src/runtime/thread_pool/worker.rs

```diff
@@ -296,13 +304,37 @@ impl Context {
         *self.core.borrow_mut() = Some(core);
 
         // Run the task
-        task.run();
-
-        // Try to take the core back
-        match self.core.borrow_mut().take() {
-            Some(core) => Ok(core),
-            None => Err(()),
-        }
+        crate::coop::budget(|| {
+            task.run();
+
+            // As long as there is budget remaining and a task exists in the
+            // `lifo_slot`, then keep running.
+            loop {
+                // Check if we still have the core. If not, the core was stolen
+                // by another worker.
+                let mut core = match self.core.borrow_mut().take() {
+                    Some(core) => core,
+                    None => return Err(()),
+                };
```

Comment:
tioli: seems like this could maybe be expressed as 
```suggestion
            // As long as there is budget remaining and a task exists in the
            // `lifo_slot`, then keep running, as long as we still have the core.
            while let Some(core) = self.core.borrow_mut().take() {